src/core/shards/imaging.cpp

```diff
@@ -165,6 +165,172 @@ struct StripAlpha {
+// Assumes that both SHVars are images. Output's data must be pre-allocated before calling this function
+// Adds SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA flag to output image
+template <typename T> void demultiplyAlpha(const SHVar &input, SHVar &output, int32_t w, int32_t h) {
+  const auto from = reinterpret_cast<T *>(input.payload.imageValue.data);
+  auto to = reinterpret_cast<T *>(output.payload.imageValue.data);
+
+  demultiplyAlpha<T>(from, to, w, h);
+
+  // remove premultiplied flag
+  output.payload.imageValue.flags &= ~SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA;
+}
+
+// Premultiplies the alpha channel in input image and writes the resulting image to a given vector of bytes
+// bytes's data must be pre-allocated before calling htis function
+// Does not add SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA flag to output image
+template <typename T> void demultiplyAlpha(const SHVar &input, std::vector<uint8_t> &bytes, int32_t w, int32_t h) {
+  const auto from = reinterpret_cast<T *>(input.payload.imageValue.data);
+  auto to = reinterpret_cast<T *>(&bytes[0]);
+
+  demultiplyAlpha<T>(from, to, w, h);
+}
+
+template <typename T> inline void demultiplyAlpha(T *from, T *to, int32_t w, int32_t h) {
+  const auto max = std::numeric_limits<T>::max(); // set to respective max value for uint8_t uint16_t and float
+
+  for (auto y = 0; y < h; y++) {
+    for (auto x = 0; x < w; x++) {
+      const auto addr = ((w * y) + x) * 4;
+      // un-premultiply RGB values
+      for (auto z = 0; z < 3; z++) {
+        to[addr + z] = static_cast<T>(static_cast<float>(from[addr + z]) / from[addr + 3] * max);
+      }
+      // copy A value
+      to[addr + 3] = from[addr + 3];
+    }
+  }
+}
+
+struct DemultiplyAlpha {
+  static SHTypesInfo inputTypes() { return CoreInfo::ImageType; }
+  static SHTypesInfo outputTypes() { return CoreInfo::ImageType; }
+
+  template <typename T> void process(const SHVar &input, SHVar &output, int32_t w, int32_t h) {
+    demultiplyAlpha<T>(input, output, w, h);
+  }
+
+  SHVar activate(SHContext *context, const SHVar &input) {
+    if (input.payload.imageValue.channels < 4)
+      return input; // nothing to do
+
+    if ((input.payload.imageValue.flags & SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA) != SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA) 
+      return input; // already straight alpha
+
+    // find number of bytes per pixel
+    auto pixsize = 1;
+    if ((input.payload.imageValue.flags & SHIMAGE_FLAGS_16BITS_INT) == SHIMAGE_FLAGS_16BITS_INT)
+      pixsize = 2;
+    else if ((input.payload.imageValue.flags & SHIMAGE_FLAGS_32BITS_FLOAT) == SHIMAGE_FLAGS_32BITS_FLOAT)
+      pixsize = 4;
+
+    int32_t w = int32_t(input.payload.imageValue.width);
+    int32_t h = int32_t(input.payload.imageValue.height);
+    
+    _bytes.resize(w * h * 4 * pixsize);
+    auto output = Var(&_bytes.front(), uint16_t(w), uint16_t(h), 4, input.payload.imageValue.flags);
+    output.version = input.version;
```

Comment:
version is used to detect changes to images so you'd want to store it in a variable inside this shard and increment it every time regardless of input version. (same for DemultiplyAlpha)

Reply:
@guusw to clarify on this, if version is stored as a variable inside the shard and incremented every time regardless of input version, do you mean that its a static variable specific to the shard? Or should each version number be unique across all images and be incremented by the use of any of the shards in `imaging.cpp`?

I think I'm a bit unsure about the use of version because at the moment, it seems that all of the shards simply assign the output version to be that of the input's version without incrementing or storing any version variable.

Reply:
Version is useful for images and bytes arrays and strings cos we do pointer comparisons sometimes but that's not fool proof, as in the end you compare just an address and a size, version is a bit like a weak hash for the content. So to know the image has changed indeed. Please search all usages of version and it should be clearer.

Reply:
If that's the case then I think the version doesn't need to be stored in a variable inside the shard, but just copied from input to output and incremented every time since version is only used as a tiebreaker. 

I have made the changes for version in commit [c02c1a6](https://github.com/fragcolor-xyz/shards/pull/682/commits/c02c1a6373528119ab82d55ccbee35232eb01bb8). The only existing shard that increments the version is `LoadImage`, but that seems to be when the values in the SHVar are first initialized, so `_output.version++` may not be initialized, so I have changed that to `_output.version = 0`, and to increment the version of the output in every shard in `Imaging.cpp`. Do let me know if my understanding conflicts with the intended behaviour for version.

Else, this should be all the requested changes addressed for this PR.